CIP-bytestring-map-builtin/README.md

```diff
@@ -0,0 +1,152 @@
+
+The only way to implement a builtin map with a (builtin) polymorphic key type would be to modify the builtins language itself. A compiler would have to only allow the construction of maps with key types which are known a priori to posess some property (such as a total ordering). The Plutus Core evaluator would have to check for this as well. The implementation itself would not be too complicated, but we must first consider how this would change the formal semantics of the builtins language.
+
+Given the above, the only immediate solution is to implement a builtin map type which is specialised to `ByteString` keys but remains polymorphic in the value type. We believe that this solution should be good enough for most practical use-cases, and maybe even surpass a fully polymorphic map type in performance.
+
+
+## Specification
+
+We propose the addition of the following builtin type: `ByteStringMap` of kind `Type -> Type`.
+- It represents a map associating keys of type `ByteString` with values of any builtin type.
+- The new builtin type should be implemented using a tree data structure allowing for logarithmic-time map operations.
+- We consider the costing size of a `ByteStringMap` to be equal to the size of its corresponding representation as an association list.
+- The binary encoding of `ByteStringMap`s is equivalent to its encoding as an association list.
+
+**Note**: Here `Type` is the universe of all builtin types, since we do not consider types formed out of applying builtin types to arbitrary types to be inhabited.
+
+We propose the following set of builtin functions to accompany the new builtin type:
+1. `insertBSMap :: forall v . ByteString -> v -> ByteStringMap v -> ByteStringMap v`
+    - it returns a map with the input key updated to the new value, silently discarding any previous value
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument, and logarithmic memory in the `ByteStringMap` argument
+2. `lookupBSMap :: forall v . ByteString -> (() -> v) -> ByteStringMap v -> v`
+    - it returns the value for a given key in the map; if the key is missing, it returns the default value of type `() -> a` 
+    - its implementation depends on, as of the time of writing, ongoing work in supporting higher-order builtin functions; see the next section for the motivation behind this choice
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument and constant memory
+3. `deleteBSMap :: forall v . ByteString -> ByteStringMap v -> ByteStringMap v`
+    - it returns a map without the key-value association identified by the input key
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument and logarithmic memory in the `ByteStringMap` argument
+4. `unionBSMap :: forall v . ByteStringMap v -> ByteStringMap v -> ByteStringMap v`
```

Comment:
@colll78 Thanks for all the pointers! Sounds like we should have `insertWithBSMap` as a builtin. With higher order function support coming soon, this makes total sense.

I'm unsure about the following:
- Should we keep `insertBSMap`
- Do we need a `unionWithBSMap` builtin, or does `insertWithBSMap` suffice
- What can we do to speed up equality checks on Values? It's not simply map equality due to potential zero-value entries. Perhaps `filterBSMap` + `equalsBSMap`?

Reply:
Map equality is cheap because we currently take advantage of `equalsData`. 

`equalsData (mergeValue ownInputValue tokensToAdd) ownOutputValue`

To handle zero-value entries `mergeValue` simply deletes them, I would expect that the union function would similarly be capable of deleting zero value entries.

Reply:
Then just a default `mergeValue` and `insertValue` (that you don't provide a function to, and only works on canonical ledger constructed `Value` as `BuiltinData`) would be just as impactful. Dealing with `Values` is the main use-case and one of the main efficiency bottlenecks in most protocols.

Reply:
For reference, perhaps one of the most universally beloved smart contract platforms in the blockchain space defines a programming paradigm "Asset oriented programming", the entire language has been built around the notion that it natively understands assets and provides native first-class features for assets. The introduction of such builtins would also vastly enhance the security of the ecosystem because the implementation of such functions in onchain code are extremely error prone, especially when you look at all the black magic optimization that libraries are using (which wouldn't be necessary with such builtins).

Reply:
@colll78 sounds interesting. Wanna open an issue in the Plutus repo about that?

However I personally prefer to be as general as possible until we're sure that that's not enough. So I'd rather investigate `unionWithBSMap`, maybe we can do that and then we can keep domain-specific logic in the user space. If that's too hard or isn't going to fly, then I [agree](https://github.com/cardano-foundation/CIPs/pull/915#discussion_r1765832949) with you.

Reply:
I think that is reasonable.
```haskell
mergeValue :: BuiltinData -> BuiltinData -> BuiltinData 
mergeValue v1 v2 = 
  let valueMap1 = listToBSMap (BI.unsafeDataAsMap v1) 
       valueMap2 = listToBSMap (BI.unsafeDataAsMap v2) 
       combinedVal = unionWithBSMap (\innerA innerB -> unionWithBsMap BI.addInteger (listToBSMap (BI.unsafeDataAsMap innerA)) (listToBSMap (BI.unsafeDataAsMap innerB))) valueMap1 valueMap2
   in toBuiltinData $ filterWithBSMap ((/=) 0) combinedVal
```

In this case `filterWithBSMap` and `equalsBSMap` are definitely required.